pkg/ingester/limiting.go

```diff
@@ -0,0 +1,239 @@
+	if err != nil {
+		return 0, 0, err
+	}
+	defer cpuR.Close()
+	// CPU time in nanoseconds
+	var cpuTime uint64
+	if _, err := fmt.Fscanf(cpuR, "%d", &cpuTime); err != nil {
+		return 0, 0, errors.Wrapf(err, "failed scanning %s", cpuPath)
+	}
+
+	return float64(cpuTime) / 1e09, memUtil, nil
+}
+
+func scanProcFS() (float64, uint64, error) {
+	p, err := procfs.Self()
+	if err != nil {
+		return 0, 0, errors.Wrap(err, "failed to get process info")
+	}
+	ps, err := p.Stat()
+	if err != nil {
+		return 0, 0, errors.Wrap(err, "failed to get process stats")
+	}
+
+	return ps.CPUTime(), uint64(ps.ResidentMemory()), nil
+}
+
+func (i *Ingester) updateResourceUtilization() {
+	if !i.cfg.UtilizationBasedLimitingEnabled {
+		return
+	}
+
+	lastUpdate := i.lastResourceUtilizationUpdate.Load()
+
+	var method string
+	cpuTime, memUtil, err := scanCgroupV2()
+	if err != nil && !os.IsNotExist(err) {
+		level.Warn(i.logger).Log("msg", "failed to get CPU and memory stats from cgroup v2", "err", err.Error())
+		return
+	}
+	if err == nil {
+		method = "cgroup v2"
+	} else {
+		// cgroup v2 not detected, try v1
+		cpuTime, memUtil, err = scanCgroupV1()
+		if err != nil && !os.IsNotExist(err) {
+			level.Warn(i.logger).Log("msg", "failed to get CPU and memory stats from cgroup v1", "err", err.Error())
+			return
+		}
+		if err == nil {
+			method = "cgroup v1"
+		} else {
+			// cgroup not detected, fall back to /proc
+			cpuTime, memUtil, err = scanProcFS()
+			if err != nil {
+				level.Warn(i.logger).Log("msg", "failed to get CPU and memory stats from /proc", "err", err.Error())
+				return
+			}
+			method = "/proc"
+		}
+	}
+
+	i.memoryUtilization.Store(memUtil)
+
+	now := time.Now().UTC()
+
+	lastCPUTime := i.lastCPUTime.Load()
+	i.lastCPUTime.Store(cpuTime)
+
+	i.lastResourceUtilizationUpdate.Store(now)
+
+	if lastUpdate.IsZero() {
+		return
+	}
+
+	cpuUtil := (cpuTime - lastCPUTime) / now.Sub(lastUpdate).Seconds()
+	i.movingAvg.Add(cpuUtil)
+	cpuA := i.movingAvg.Value()
+	i.cpuUtilization.Store(cpuA)
+
+	level.Debug(i.logger).Log("msg", "process resource utilization", "method", method, "memory_utilization", memUtil,
+		"smoothed_cpu_utilization", cpuA, "raw_cpu_utilization", cpuUtil)
+}
+
+// checkReadOverloaded checks whether the ingester read path is overloaded wrt. CPU and/or memory.
+func (i *Ingester) checkReadOverloaded() error {
+	if !i.cfg.UtilizationBasedLimitingEnabled {
+		return nil
+	}
+
+	memUtil := i.memoryUtilization.Load()
+	cpuUtil := i.cpuUtilization.Load()
+	lastUpdate := i.lastResourceUtilizationUpdate.Load()
+	if lastUpdate.IsZero() {
+		return nil
+	}
+
+	memPercent := 100 * (float64(memUtil) / float64(i.readPathMemoryThreshold))
+	cpuPercent := 100 * (cpuUtil / i.readPathCPUThreshold)
+
+	var reason string
+	if memPercent >= 100 {
+		reason = "memory"
+	} else if cpuPercent >= 100 {
+		reason = "cpu"
+	}
+	if reason != "" {
+		level.Debug(i.logger).Log("msg", "read path resource utilization based limiting", "reason", reason,
```

Comment:
If you accept my suggestion above to pre-compute whether the read-path is limited directly in `updateResourceUtilization()`, then I would log this only once when the limiting flag switches (enabled/disabled) with an info level (so that it's actually visible) rather than as a debug for each request (in prod it's unlikely you have debug level enabled).

Reply:
I've done as you ask :) PTAL.